File: pyfiles/v1/gather_metric.py
import numpy as np
import os
from netCDF4 import Dataset

def gather_metric(metricinfo,varnm,metricdirvar,metric_dir):

    '''
    gather metric data
    Author: Yuanhang Huo
    Email:  yhhuo@mail.ustc.edu.cn
    Date:   2021.06.24
    '''

    if os.path.exists(metric_dir):

        metricprefix='metric'
        nthd=len(metricinfo)

        for n in range(nthd):

            n_i=metricinfo[n]['thisid'][0]
            n_j=metricinfo[n]['thisid'][1]

            i1=metricinfo[n]['indxs'][0]
            j1=metricinfo[n]['indxs'][1]
            k1=metricinfo[n]['indxs'][2]
            
            i2=metricinfo[n]['indxe'][0]
            j2=metricinfo[n]['indxe'][1]
            k2=metricinfo[n]['indxe'][2]

            subs=metricinfo[n]['wsubs']
            subc=metricinfo[n]['wsubc']
            subt=metricinfo[n]['wsubt']

            fnm_metric=metric_dir + "/" + metricprefix  + "_px" + str(n_i) + \
                     "_py" + str(n_j) + ".nc"

            if os.path.exists(fnm_metric):

                metricnc=Dataset(fnm_metric)

                subs=subs[::-1]
                subc=subc[::-1]
                subt=subt[::-1]
                
                vblk=np.array(metricnc.variables[varnm][\
                        subs[0]-1:subs[0]+subc[0]*subt[0]-1:subt[0],\
                        subs[1]-1:subs[1]+subc[1]*subt[1]-1:subt[1],\
                        subs[2]-1:subs[2]+subc[2]*subt[2]-1:subt[2]],\
                        dtype=np.float64)

                if n == 0:
                    vold=vblk

                vnew=np.empty((max(k2,vold.shape[0]),max(j2,vold.shape[1]),max(i2,vold.shape[2])))

                vnew[:vold.shape[0],:vold.shape[1],:vold.shape[2]]=vold

                vnew[k1-1:k2,j1-1:j2,i1-1:i2]=vblk

                vold=vnew

            else:
                print("Error in 'gather_metric': File " + "'" + fnm_metric + "'" + \
                      " does NOT exist.")

        vnew=vnew.transpose(2,1,0)

        return(vnew)

    else:
        print("Error in 'gather_metric': Directory " + "'" + metric_dir + "'" + \
              " does NOT exist.")

File: mfiles/gather_metric.m
function [V,varargout] = gather_metric(metricinfo,varargin)

% gather metric data
% Author:   Yuanhang Huo
% Email:    yhhuo@mail.ustc.edu.cn
% Date:     2021.06.06

nargs=nargin-1;

n=1;
while n<=nargs
    
    if numel(varargin{n})==1 | ~isnumeric(varargin{n})
        switch varargin{n}
            case {'jac','xi_x','xi_y','xi_z',...
                  'eta_x','eta_y','eta_z',...
                  'zeta_x','zeta_y','zeta_z'}
                varnm=varargin{n};
            case 'metricdir'
                metric_dir=varargin{n+1}; n=n+1;
        end
    end
    
    n=n+1;
    
end

% check path exists
if ~ exist(metric_dir,'dir')
   error([mfilename ': directory ' metric_dir ' does not exist']);
end

% load
metricprefix='metric';
nthd=length(metricinfo);
for n=1:nthd
    
    n_i=metricinfo(n).thisid(1); n_j=metricinfo(n).thisid(2);
    i1=metricinfo(n).indxs(1); j1=metricinfo(n).indxs(2); k1=metricinfo(n).indxs(3);
    i2=metricinfo(n).indxe(1); j2=metricinfo(n).indxe(2); k2=metricinfo(n).indxe(3);
    subs=metricinfo(n).wsubs; subc=metricinfo(n).wsubc; subt=metricinfo(n).wsubt;
    fnm_metric=[metric_dir,'/',metricprefix,'_px',num2str(n_i),'_py',num2str(n_j),'.nc'];
    
    if ~ exist(fnm_metric,'file')
       error([mfilename ': file ' fnm_metric 'does not exist']);
    end

    V(k1:k2,j1:j2,i1:i2)=nc_varget(fnm_metric,varnm, ...
                            fliplr(subs)-1,fliplr(subc),fliplr(subt));
                               
end

V=double(permute(V,[3,2,1]));

% pack varargout
if nargout>=2
    varargout(1)={varnm};
end


end